### PR TITLE
✨Source Outreach: Migrate Python CDK to Low-code CDK

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
*Migrating Source Outreach to Low-Code CDK*
Closes airbytehq/airbyte-internal-issues#6916

## How
*Developed using (Configuration Based Source) low-code CDK*

## Recommended reading order
1. `spec.yaml`
2. `manifest.yaml`
3. `schemas/*`

## Tests

<details open><summary><strong>Integration & Acceptance</strong></summary>

 - Test successful with incremental syncs

image

</details>

## 🚨 User Impact 🚨
- Migration to low-code

### CODE REVIEW

**Comment by btkcodedev** [comment]
Inline schema :white_check_mark: 
Poetry lock file ✅
Breaking change referred in metadata :white_check_mark: 
Add migration.md docs file :white_check_mark: 
Docs updated ✅
Formatted :white_check_mark: 


`airbyte-ci test` - All pass
image

**Comment by btkcodedev** [comment]
Marking PR as ready to merge

**Comment by girarda** [comment]
@btkcodedev can you run `airbyte-ci format fix all` to fix the formatting issues?

**Comment by btkcodedev** [comment]
Done @girarda !!

**Review by lazebnyi — CHANGES_REQUESTED** [review]
During testing with regressions tests found: 

- Check connection can not be passed as in new version have added unnecessary scopes list
- Catalog not the same, as changed field type in sequence_steps_schema to string and integer at the same time. Need change to string and add type transformer
- Streams sequence_states and prospects return less records with state sync than expected 

Is it expected for fields to be removed from the schema for certain streams? : 

- accounts
Item root['properties']['assignedTeams'] removed from dictionary.
Item root['properties']['batches'] removed from dictionary.
Item root['properties']['favorites'] removed from dictionary.
Item root['properties']['prospects'] removed from dictionary.
Item root['properties']['tasks'] removed from dictionary.
Item root['properties']['updater'] removed from dictionary.
Item root['required'][0] removed from iterable.
Item root['required'][2] removed from iterable.
Item root['required'][6] removed from iterable.
Item root['required'][162] removed from iterable.
Item root['required'][175] removed from iterable.
Item root['required'][178] removed from iterable.
Item root['required'][182] removed from iterable.

-  users
Item root['properties']['batches'] removed from dictionary.
Item root['properties']['contentCategories'] removed from dictionary.
Item root['properties']['contentCategoryOwnerships'] removed from dictionary.
Item root['properties']['favorites'] removed from dictionary.
Item root['properties']['jobRole'] removed from dictionary.
Item root['properties']['recipients'] removed from dictionary.
Item root['properties']['voicemailPrompts'] removed from dictionary.
Item root['required'][3] removed from iterable.
Item root['required'][9] removed from iterable.
Item root['required'][12] removed from iterable.
Item root['required'][13] removed from iterable.
Item root['required'][16] removed from iterable.
Item root['required'][31] removed from iterable.
Item root['required'][43] removed from iterable.
Item root['required'][73] removed from iterable.
Item root['required'][75] removed from iterable.
Item root['required'][79] removed from iterable.
Item root['required'][88] removed from iterable.
Item root['required'][102] removed from iterable.
Item root['required'][106] removed from iterable.
Item root['required'][114] removed from iterable.
Item root['required'][120] removed from iterable.

- mailings
Item root['properties']['attachments'] removed from dictionary.
Item root['properties']['followUpSequence'] removed from dictionary.
Item root['required'][0] removed from iterable.
Item root['required'][6] removed from iterable.
Item root['required'][14] removed from iterable.
Item root['required'][35] removed from iterable.
Item root['required'][39] removed from iterable.
Item root['required'][46] removed from iterable.
Item root['required'][49] removed from iterable.
Item root['required'][50] removed from iterable.
Item root['required'][51] removed from iterable.
Item root['required'][55] removed from iterable.
Item root['required'][56] removed from iterable.
Item root['required'][57] removed from iterable.

- calls
Item root['properties']['batches'] removed from dictionary.
Item root['properties']['opportunity'] removed from dictionary.
Item root['required'][1] removed from iterable.
Item root['required'][2] removed from iterable.
Item root['required'][3] removed from iterable.
Item root['required'][12] removed from iterable.
Item root['required'][13] removed from iterable.
Item root['required'][16] removed from iterable.
Item root['required'][17] removed from iterable.
Item root['required'][21] removed from iterable.
Item root['required'][23] removed from iterable.
Item root['required'][24] removed from iterable.
Item root['required'][29] removed from iterable.

- stages
Item root['properties']['creator'] removed from dictionary.
Item root['properties']['prospects'] removed from dictionary.
Item root['properties']['updater'] removed from dictionary.
Item root['required'][2] removed from iterable.
Item root['required'][6] removed from iterable.
Item root['required'][8] removed from iterable.

- prospects
Item root['properties']['assignedTeams'] removed from dictionary.
Item root['properties']['batches'] removed from dictionary.
Item root['properties']['calls'] removed from dictionary.
Item root['properties']['mailings'] removed from dictionary.
Item root['properties']['opportunities'] removed from dictionary.
Item root['properties']['opportunityProspectRoles'] removed from dictionary.
Item root['properties']['persona'] removed from dictionary.
Item root['properties']['sequenceStates'] removed from dictionary.
Item root['properties']['tasks'] removed from dictionary.
Item root['properties']['trashedByAccount'] removed from dictionary.
Item root['required'][2] removed from iterable.
Item root['required'][11] removed from iterable.
Item root['required'][14] removed from iterable.
Item root['required'][16] removed from iterable.
Item root['required'][34] removed from iterable.
Item root['required'][186] removed from iterable.
Item root['required'][201] removed from iterable.
Item root['required'][216] removed from iterable.
Item root['required'][223] removed from iterable.
Item root['required'][224] removed from iterable.
Item root['required'][229] removed from iterable.
Item root['required'][233] removed from iterable.
Item root['required'][240] removed from iterable.
Item root['required'][249] removed from iterable.
Item root['required'][252] removed from iterable.
Item root['required'][259] removed from iterable.
Item root['required'][263] removed from iterable.

- sequence_states
Item root['properties']['calls'] removed from dictionary.
Item root['properties']['mailings'] removed from dictionary.
Item root['properties']['opportunity'] removed from dictionary.
Item root['properties']['tasks'] removed from dictionary.
Item root['required'][2] removed from iterable.
Item root['required'][3] removed from iterable.
Item root['required'][7] removed from iterable.
Item root['required'][10] removed from iterable.
Item root['required'][16] removed from iterable.
Item root['required'][20] removed from iterable.
Item root['required'][30] removed from iterable.
Item root['required'][33] removed from iterable.

- tasks
Item root['properties']['account'] removed from dictionary.
Item root['properties']['calls'] removed from dictionary.
Item root['properties']['creator'] removed from dictionary.
Item root['properties']['mailings'] removed from dictionary.
Item root['properties']['opportunity'] removed from dictionary.
Item root['properties']['sequenceStateSequenceStepOverrides'] removed from dictionary.
Item root['properties']['sequenceStateStartingTemplate'] removed from dictionary.
Item root['properties']['sequenceStepOverrideTemplates'] removed from dictionary.
Item root['properties']['template'] removed from dictionary.
Item root['required'][0] removed from iterable.
Item root['required'][3] removed from iterable.
Item root['required'][4] removed from iterable.
Item root['required'][8] removed from iterable.
Item root['required'][10] removed from iterable.
Item root['required'][11] removed from iterable.
Item root['required'][14] removed from iterable.
Item root['required'][15] removed from iterable.
Item root['required'][17] removed from iterable.
Item root['required'][24] removed from iterable.
Item root['required'][25] removed from iterable.
Item root['required'][31] removed from iterable.
Item root['required'][32] removed from iterable.
Item root['required'][34] removed from iterable.
Item root['required'][35] removed from iterable.
Item root['required'][36] removed from iterable.
Item root['required'][43] removed from iterable.

**Comment by btkcodedev** [comment]
@lazebnyi, 
1) those schemas looks exactly the same as previous version json, Could you please look again where is the fault, the connection looks good in local and also in Airbyte ci to, 

2) There is a schema change from backend in sequence_steps_schema where changed its ['properties']['creator']['items']['type'] field to include integers, some of them are integers and some of them are strings

3) As the scope are not defined in the previous version and is hardcoded, user may need to access the resources, so for safety I've included all the scopes necessary for the streams.

**Comment by lazebnyi** [comment]
1. Ok, I'll check and get back to you. 
2. In that way we need use string in schema and use TypeTransformer for integers 
3. However, having all scopes is not necessary to sync a specific stream. By default, the availability check will filter the stream if it's not accessible to the user.

**Comment by btkcodedev** [comment]
@lazebnyi Any updates🙂

**Comment by lazebnyi** [comment]
Please check my comment above. 

You need update schema and use TypeTransformer and delete scopes list in the code as it is not require to have all scopes to start sync for some special streams.

**Comment by btkcodedev** [comment]
@lazebnyi, The Schema look good, and it already includes all the mentioned fields in https://github.com/airbytehq/airbyte/pull/36954#pullrequestreview-2002425832

There is no need to TypeTransformation not needed as it includes only integers instead of string,
image

Scope has been reset according to the given LastPass notes

The connection is failing as the refresh token is expired I think, The creds need to be refreshed

**Comment by btkcodedev** [comment]
(Closed and reopened to run CI again)

**Comment by btkcodedev** [comment]
@lazebnyi
Scopes has been reset to the ones in LastPass,
Connection check and all other checks are passing in the CI,
Schema looks same as mentioned, 
Please continue the merge process

**Comment by btkcodedev** [comment]
CI Passed perfectly
image

File: airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,5755 @@
...
+        items:
+          type: integer
+      creator:
+        type:
+          - "null"
+          - array
+        items:
+          type:
+            - string
+            - integer
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml:4283]: Leave one of data type and use transformer to update field data to schema filed type

File: airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml [hunk 2]
```diff
@@ -0,0 +1,5755 @@
...
+      page[size]: "1000"
+    request_headers: {}
+    authenticator:
+      type: OAuthAuthenticator
+      client_id: "{{ config['client_id'] }}"
+      client_secret: "{{ config['client_secret'] }}"
+      refresh_token: "{{ config['refresh_token'] }}"
+      token_refresh_endpoint: https://api.outreach.io/oauth/token
+      grant_type: refresh_token
+      scopes:
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml:197]: Please delete the scopes. Having all scopes is not necessary to sync a specific stream. By default, the availability check should filter the stream if it's not accessible to the user.

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml (hunk 2 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml:197]: The scopes are given according to the credential, which are the scopes to be deleted?

Comment (btkcodedev) [reply airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml:197]: @lazebnyi

Comment (lazebnyi) [reply airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml:197]: All, as I said if user what to sync one stream he no need grant all scopes.

File: airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml (hunk 1 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml:4245]: There is a mismatch, there are not strings In that, only integers, still a breaking change, but no need to transformer I think, Corrected!

File: airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml (hunk 2 above)
Comment (btkcodedev) [reply airbyte-integrations/connectors/source-outreach/source_outreach/manifest.yaml]: Done!

**Comment by btkcodedev** [comment]
@lazebnyi, Resolved comments and CI green :white_check_mark: 
Ready to merge :bow:

**Review by lazebnyi — APPROVED** [review]
LGTM
@katmarkham could you check breaking changes docs before merge

**Comment by btkcodedev** [comment]
@katmarkham, Docs checked, Ready for the approval :bow:

**Review by katmarkham — APPROVED** [review]

**Pull request merged**